kinto/core/resource/__init__.py

```diff
@@ -1041,21 +1042,35 @@ def _extract_pagination_rules_from_token(self, limit, sorting):
         filters = []
         offset = 0
         if token:
+            error_msg = None
             try:
                 tokeninfo = json.loads(decode64(token))
                 if not isinstance(tokeninfo, dict):
                     raise ValueError()
                 last_record = tokeninfo['last_record']
                 offset = tokeninfo['offset']
+                nonce = tokeninfo['nonce']
             except (ValueError, KeyError, TypeError):
                 error_msg = '_token has invalid content'
+
+            # We don't want pagination tokens to be reused several times (#1171).
+            # The cache backend is used to keep track of "nonces".
+            if self.request.method.lower() == "delete" and error_msg is None:
+                registry = self.request.registry
+                if registry.cache.get(nonce) is None:
+                    error_msg = '_token was already used or has expired.'
+                else:
+                    registry.cache.delete(nonce)
```

Comment:
It's still better than what we have and I'd be happy to merge it if there's a FIXME comment explaining the race and why we can't fix it.